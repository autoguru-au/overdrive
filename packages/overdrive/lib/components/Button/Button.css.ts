import { style, styleVariants } from '@vanilla-extract/css';

import { vars } from '../../themes/base/vars.css';

export const root = style({
	transitionDelay: '0s',
	transitionTimingFunction: vars.animation.easing.standard,
	transitionDuration: '0.1s',
	transitionProperty:
		'color, background-color, border-color, box-shadow, transform',
	transform: 'translate3d(0, 0, 1) scale(1)',
	willChange: 'transform',
});

export const body = style({
	display: 'grid',
	gridAutoFlow: 'column dense',
	gridGap: vars.space['1'],
});

export const hiddenContent = style({
	visibility: 'hidden',
});
export const loading = style({
	cursor: 'not-allowed',
});

export const enabled = style({
	cursor: 'pointer',
});

export const disabled = style({
	cursor: 'not-allowed',
	opacity: '0.3',
});

export const spinner = style({
	margin: '0 auto',
});

const smallHeight = '36px';

export const size = {
	small: styleVariants({
		default: {
			minWidth: vars.space['8'],
			height: smallHeight,
			gridGap: vars.space['1'],
		},
		rounded: {
			minWidth: smallHeight,
		},
		iconOnly: {
			width: smallHeight,
		},
	}),
	medium: styleVariants({
		default: {
			minWidth: vars.space['9'],
			height: vars.space['8'],
			gridGap: vars.space['2'],
		},
		rounded: {
			minWidth: vars.space['8'],
		},
		iconOnly: {
			width: vars.space['8'],
		},
	}),
};
export const variant = {
	primary: style({
		color: vars.colours.intent.primary.foreground,
		backgroundColor: vars.colours.intent.primary.background.standard,
		boxShadow: `inset 0 0 0 1px ${vars.colours.intent.primary.border}, ${vars.elevation['2']}`,
	}),
	secondary: style({
		color: vars.colours.intent.secondary.foreground,
		backgroundColor: vars.colours.intent.secondary.background.standard,
		boxShadow: `inset 0 0 0 1px ${vars.colours.intent.secondary.border}, ${vars.elevation['2']}`,
	}),
	danger: style({
		backgroundColor: vars.colours.intent.danger.background.standard,
		color: vars.colours.intent.danger.foreground,
		boxShadow: `inset 0 0 0 1px ${vars.colours.intent.danger.border}, ${vars.elevation['2']}`,
	}),
};

export const defaultStates = {
	primary: style({
		':hover': {
			color: vars.colours.intent.primary.foreground,
			backgroundColor: vars.colours.intent.primary.background.strong,
			boxShadow: `inset 0 0 0 1px ${vars.colours.intent.primary.border}, ${vars.elevation['3']}`,
		},
		':active': {
			transform: 'scale(0.97)',
			boxShadow: `inset 0 0 0 1px ${vars.colours.intent.primary.border}, ${vars.elevation['1']}`,
			color: vars.colours.intent.primary.foreground,
			backgroundColor: vars.colours.intent.primary.background.strong,
		},
	}),
	secondary: style({
		':hover': {
			color: vars.colours.intent.secondary.foreground,
			backgroundColor: vars.colours.intent.secondary.background.strong,
			boxShadow: `inset 0 0 0 1px ${vars.colours.intent.secondary.border}, ${vars.elevation['3']}`,
		},
		':active': {
			transform: 'scale(0.97)',
			color: vars.colours.intent.secondary.foreground,
			backgroundColor: vars.colours.intent.secondary.background.strong,
			boxShadow: `inset 0 0 0 1px ${vars.colours.intent.secondary.border}, ${vars.elevation['1']}`,
		},
	}),
	danger: style({
		':hover': {
			color: vars.colours.intent.danger.foreground,
			backgroundColor: vars.colours.intent.danger.background.strong,
			boxShadow: `inset 0 0 0 1px ${vars.colours.intent.danger.border}, ${vars.elevation['3']}`,
		},
		':active': {
			transform: 'scale(0.97)',
			color: vars.colours.intent.danger.foreground,
			backgroundColor: vars.colours.intent.danger.background.strong,
			boxShadow: `inset 0 0 0 1px ${vars.colours.intent.danger.border}, ${vars.elevation['1']}`,
		},
	}),
};

export const minimal = {
	defaults: style({
		color: vars.typography.colour.neutral,
	}),
	noneRounded: style({
		minWidth: '50px',
	}),
};

export const minimalStates = {
	primary: style({
		':hover': {
			color: vars.colours.intent.primary.background.strong,
			backgroundColor: vars.colours.intent.primary.background.mild,
			boxShadow: 'none',
		},
		':active': {
			color: vars.colours.intent.primary.background.strong,
			backgroundColor: vars.colours.intent.primary.background.mild,
			boxShadow: 'none',
		},
	}),
	secondary: style({
		':hover': {
			color: vars.typography.colour.secondary,
			backgroundColor: vars.colours.intent.secondary.background.strong,
			boxShadow: 'none',
		},
		':active': {
			color: vars.typography.colour.secondary,
			backgroundColor: vars.colours.intent.secondary.background.strong,
			boxShadow: 'none',
		},
	}),
	danger: style({
		':hover': {
			color: vars.colours.intent.danger.background.strong,
			backgroundColor: vars.colours.intent.danger.background.mild,
			boxShadow: 'none',
		},
		':active': {
			color: vars.colours.intent.danger.background.strong,
			backgroundColor: vars.colours.intent.danger.background.mild,
			boxShadow: 'none',
		},
	}),
};
