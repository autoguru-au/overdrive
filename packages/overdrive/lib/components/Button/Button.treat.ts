import { style, styleMap } from 'treat';

import { shadedColour } from '../../themes/helpers';

export const root = style(({ animation }) => ({
	transitionDelay: '0s',
	transitionTimingFunction: animation.easing.standard,
	transitionDuration: '0.1s',
	transitionProperty:
		'color, background-color, border-color, box-shadow, transform',
	transform: 'translate3d(0, 0, 1) scale(1)',
	willChange: 'transform',
}));

export const body = style(({ space }) => ({
	display: 'grid',
	gridAutoFlow: 'column dense',
	gridGap: space['1'],
}));

export const loading = style({
	cursor: 'not-allowed',
});

export const enabled = style({
	cursor: 'pointer',
});

export const disabled = style({
	cursor: 'not-allowed',
	opacity: '0.3',
});

export const spinner = style({
	margin: '0 auto',
});

const smallHeight = '36px';

export const size = {
	small: styleMap(({ space }) => ({
		default: {
			minWidth: space['8'],
			height: smallHeight,
			gridGap: space['1'],
		},
		rounded: {
			minWidth: smallHeight,
		},
		iconOnly: {
			width: smallHeight,
		},
	})),
	medium: styleMap(({ space }) => ({
		default: {
			minWidth: space['9'],
			height: space['8'],
			gridGap: space['2'],
		},
		rounded: {
			minWidth: space['8'],
		},
		iconOnly: {
			width: space['8'],
		},
	})),
};
export const variant = {
	primary: style(({ colours, shadeIntensity, isDark, elevation }) => ({
		color: colours.intent.primary.foreground,
		backgroundColor: colours.intent.primary.background,
		boxShadow: `inset 0 0 0 1px ${shadedColour(
			colours.intent.primary.background,
			shadeIntensity.slight,
			'backward',
			isDark,
		)}, ${elevation['2']}`,
	})),
	secondary: style(({ colours, elevation }) => ({
		color: colours.intent.secondary.foreground,
		backgroundColor: colours.intent.secondary.background,
		boxShadow: `inset 0 0 0 1px ${colours.intent.secondary.foreground}, ${elevation['2']}`,
	})),
	danger: style(({ colours, shadeIntensity, isDark, elevation }) => ({
		backgroundColor: colours.intent.danger.background,
		color: colours.intent.danger.foreground,
		boxShadow: `inset 0 0 0 1px ${shadedColour(
			colours.intent.danger.background,
			shadeIntensity.slight,
			'backward',
			isDark,
		)}, ${elevation['2']}`,
	})),
};

export const defaultStates = {
	primary: style(({ colours, shadeIntensity, isDark, elevation }) => ({
		':hover': {
			color: colours.intent.primary.foreground,
			backgroundColor: shadedColour(
				colours.intent.primary.background,
				shadeIntensity.slight,
				'backward',
				isDark,
			),
			boxShadow: `inset 0 0 0 1px ${shadedColour(
				colours.intent.primary.background,
				shadeIntensity.slight,
				'backward',
				isDark,
			)}, ${elevation['3']}`,
		},
		':active': {
			transform: 'scale(0.97)',
			boxShadow: `inset 0 0 0 1px ${shadedColour(
				colours.intent.primary.background,
				shadeIntensity.medium,
				'backward',
				isDark,
			)}, ${elevation['1']}`,
			color: colours.intent.primary.foreground,
			backgroundColor: shadedColour(
				colours.intent.primary.background,
				shadeIntensity.medium,
				'backward',
				isDark,
			),
		},
	})),
	secondary: style(({ colours, isDark, shadeIntensity, elevation }) => ({
		':hover': {
			color: colours.intent.secondary.foreground,
			backgroundColor: shadedColour(
				colours.intent.secondary.background,
				shadeIntensity.slight,
				'backward',
				isDark,
			),
			boxShadow: `inset 0 0 0 1px ${shadedColour(
				colours.intent.secondary.background,
				shadeIntensity.medium,
				'backward',
				isDark,
			)}, ${elevation['3']}`,
		},
		':active': {
			transform: 'scale(0.97)',
			color: colours.intent.secondary.foreground,
			backgroundColor: shadedColour(
				colours.intent.secondary.background,
				shadeIntensity.medium,
				'backward',
				isDark,
			),
			boxShadow: `inset 0 0 0 1px ${shadedColour(
				colours.intent.secondary.background,
				shadeIntensity.intense,
				'backward',
				isDark,
			)}, ${elevation['1']}`,
		},
	})),
	danger: style(({ colours, shadeIntensity, isDark, elevation }) => ({
		':hover': {
			color: colours.intent.danger.foreground,
			backgroundColor: shadedColour(
				colours.intent.danger.background,
				shadeIntensity.slight,
				'backward',
				isDark,
			),
			boxShadow: `inset 0 0 0 1px ${shadedColour(
				colours.intent.danger.background,
				shadeIntensity.medium,
				'backward',
				isDark,
			)}, ${elevation['3']}`,
		},
		':active': {
			transform: 'scale(0.97)',
			color: colours.intent.danger.foreground,
			backgroundColor: shadedColour(
				colours.intent.danger.background,
				shadeIntensity.medium,
				'backward',
				isDark,
			),
			boxShadow: `inset 0 0 0 1px ${shadedColour(
				colours.intent.danger.background,
				shadeIntensity.intense,
				'backward',
				isDark,
			)}, ${elevation['1']}`,
		},
	})),
};

export const minimal = {
	defaults: style(({ typography }) => ({
		color: typography.colour.neutral,
	})),
	noneRounded: style({
		minWidth: '50px',
	}),
};

export const minimalStates = {
	primary: style(({ colours, isDark, shadeIntensity, transparency }) => ({
		':hover': {
			color: shadedColour(
				colours.intent.primary.background,
				shadeIntensity.medium,
				'backward',
				isDark,
			),
			backgroundColor: shadedColour(
				colours.intent.primary.background,
				shadeIntensity.intense,
				'forward',
				isDark,
				transparency.intense,
			),
			boxShadow: 'none',
		},
		':active': {
			color: shadedColour(
				colours.intent.primary.background,
				shadeIntensity.medium,
				'backward',
				isDark,
			),
			backgroundColor: shadedColour(
				colours.intent.primary.background,
				shadeIntensity.intense,
				'backward',
				isDark,
				transparency.intense,
			),
			boxShadow: 'none',
		},
	})),
	secondary: style(
		({ colours, typography, shadeIntensity, isDark, transparency }) => ({
			':hover': {
				color: shadedColour(
					typography.colour.secondary,
					shadeIntensity.medium,
					'backward',
					isDark,
				),
				backgroundColor: shadedColour(
					colours.intent.secondary.foreground,
					shadeIntensity.intense,
					'forward',
					isDark,
					transparency.intense,
				),
				boxShadow: 'none',
			},
			':active': {
				color: shadedColour(
					typography.colour.secondary,
					shadeIntensity.medium,
					'backward',
					isDark,
				),
				backgroundColor: shadedColour(
					colours.intent.secondary.foreground,
					shadeIntensity.medium,
					'backward',
					isDark,
					transparency.intense,
				),
				boxShadow: 'none',
			},
		}),
	),
	danger: style(({ colours, shadeIntensity, isDark, transparency }) => ({
		':hover': {
			color: shadedColour(
				colours.intent.danger.background,
				shadeIntensity.medium,
				'backward',
				isDark,
			),
			backgroundColor: shadedColour(
				colours.intent.danger.background,
				shadeIntensity.intense,
				'forward',
				isDark,
				transparency.intense,
			),
			boxShadow: 'none',
		},
		':active': {
			color: shadedColour(
				colours.intent.danger.background,
				shadeIntensity.medium,
				'backward',
				isDark,
			),
			backgroundColor: shadedColour(
				colours.intent.danger.background,
				shadeIntensity.medium,
				'backward',
				isDark,
				transparency.intense,
			),
			boxShadow: 'none',
		},
	})),
};
