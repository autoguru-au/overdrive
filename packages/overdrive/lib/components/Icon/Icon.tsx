import type { IconType } from '@autoguru/icons';
import type { FunctionComponent, ReactElement, SVGAttributes } from 'react';
import * as React from 'react';
import { cloneElement } from 'react';

import { resolveResponsiveStyle } from '../../utils/resolveResponsiveProps';
import { ResponsiveProp } from '../../utils/responsiveProps.css';
import type { BoxStyleProps } from '../Box';
import { Box, useBoxStyles } from '../Box';

import * as styles from './Icon.css';

export interface Props {
	display?: Extract<BoxStyleProps['display'], 'block' | 'inlineBlock'>;
	className?: string;
	size?: ResponsiveProp<keyof typeof styles.size | string>;
	icon: IconType | ReactElement<SVGAttributes<SVGElement>, 'svg'>;
}

export const Icon: FunctionComponent<Props> = ({
	className = '',
	icon,
	size = 'small',
	display = 'block',
}) => (
	<Box
		is="i"
		display={display}
		className={[resolveResponsiveStyle(size, styles.size), className]}
		role="presentation">
		{cloneElement(icon, {
			className: useBoxStyles({
				is: 'svg',
				display: 'block',
				width: 'full',
				height: 'full',
			}),
		})}
	</Box>
);
