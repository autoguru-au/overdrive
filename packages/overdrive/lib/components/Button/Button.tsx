import { IconType } from '@autoguru/icons';
import clsx from 'clsx';
import * as React from 'react';
import {
	AriaAttributes,
	ButtonHTMLAttributes,
	cloneElement,
	ComponentProps,
	createElement,
	ElementType,
	forwardRef,
	isValidElement,
	ReactElement,
	useMemo,
} from 'react';
import { useStyles } from 'react-treat';

import { Box, useBoxStyles } from '../Box';
import { Icon } from '../Icon';
import { ProgressSpinner } from '../ProgressSpinner';
import { useTextStyles } from '../Text';

import * as styleRefs from './Button.treat';

type ButtonPrimitive = ButtonHTMLAttributes<HTMLButtonElement>;

type AllowedChildren = string | IconType;

export interface Props
	extends Pick<ButtonPrimitive, 'id' | 'onClick' | 'type' | 'className'>,
		Pick<AriaAttributes, 'aria-label'> {
	children: AllowedChildren | AllowedChildren[];
	disabled?: boolean;
	is?: ElementType | ReactElement;
	isLoading?: boolean;
	isFullWidth?: boolean;
	minimal?: boolean;
	rounded?: boolean;
	size?: keyof typeof styleRefs.size;
	variant?: keyof typeof styleRefs.variant;
}

const getSpinnerColour: (
	variant: keyof typeof styleRefs.variant,
	minimal: boolean,
) => ComponentProps<typeof ProgressSpinner>['colour'] = (variant, minimal) =>
	minimal || variant === 'secondary' ? 'secondary' : 'light';

const getBorderRadius: (
	rounded: boolean,
) => ComponentProps<typeof Box>['borderRadius'] = (rounded) =>
	rounded ? 'pill' : '1';

const getPadding: (
	size: keyof typeof styleRefs.size,
	loading: boolean,
) => ComponentProps<typeof Box>['paddingX'] = (size, loading) => {
	if (loading) return 'none';
	return size === 'small' ? '3' : '4';
};

export const Button = forwardRef<HTMLButtonElement, Props>(
	(
		{
			children,
			className = '',
			disabled = false,
			id,
			is: Component = 'button',
			isLoading = false,
			isFullWidth = false,
			minimal = false,
			onClick,
			rounded = false,
			size = 'medium',
			type = 'button',
			variant = 'secondary',
			'aria-label': ariaLabel,
		},
		ref,
	) => {
		const styles = useStyles(styleRefs);
		const { isSingleIconChild, props: maybeIconProps } = useMemo(() => {
			const maybeIcon =
				isValidElement(children) && children.type === Icon;
			const maybeProps = children as ReactElement<
				ComponentProps<typeof Icon>
			>;

			return maybeIcon
				? {
						isSingleIconChild: true,
						props: maybeProps.props,
				  }
				: { isSingleIconChild: false };
		}, [children]);

		const props: Partial<ButtonPrimitive> & { ref: typeof ref } = {
			type: Component === 'button' ? (type as any) : undefined,
			id,
			onClick,
			disabled: disabled || isLoading,
			'aria-label': ariaLabel,
			className: clsx(
				useBoxStyles({
					is: typeof Component === 'string' ? Component : undefined,
					display: 'inlineBlock',
					overflow: 'hidden',
					borderRadius: getBorderRadius(rounded),
					textAlign: 'center',
					borderWidth: 'none',
					paddingY: 'none',
					paddingX: getPadding(size, isLoading),
					width: isFullWidth ? 'full' : void 0,
				}),
				styles.root,
				useTextStyles({
					colour: 'white',
					fontWeight: 'semiBold',
					size: size === 'medium' ? '4' : '3',
				}),
				getButtonStates(
					styles,
					variant,
					disabled || isLoading,
					minimal,
					rounded,
				),
				!minimal && styles.variant[variant],
				getButtonSize(
					styles,
					size,
					rounded,
					isSingleIconChild && !isLoading,
				),
				{
					[styles.disabled]: disabled,
					[styles.enabled]: !disabled && !isLoading,
					[styles.loading]: isLoading,
				},
				className,
			),
			ref,
		};

		const child = isLoading ? (
			<ProgressSpinner
				className={styles.spinner}
				colour={getSpinnerColour(variant, minimal)}
			/>
		) : (
			<Box
				alignItems="center"
				justifyContent="center"
				height="full"
				className={styles.body}>
				{isSingleIconChild && maybeIconProps ? (
					<Icon
						size={
							maybeIconProps.size ?? size === 'small'
								? 'small'
								: 'medium'
						}
						{...maybeIconProps}
					/>
				) : (
					children
				)}
			</Box>
		);

		return isValidElement(Component)
			? // @ts-ignore
			  cloneElement(Component, { ref, ...props }, child)
			: // @ts-ignore
			  createElement(Component, { ref, ...props }, child);
	},
);

const getButtonStates: (
	styles: typeof styleRefs,
	variant: keyof typeof styleRefs.variant,
	disabled: boolean,
	minimal: boolean,
	rounded: boolean,
) => string = (styles, variant, disabled, minimal, rounded) => {
	if (disabled)
		return minimal
			? clsx(styles.minimal.defaults, {
					[styles.minimal.noneRounded]: !rounded,
			  })
			: '';

	if (minimal)
		return clsx(styles.minimal.defaults, styles.minimalStates[variant], {
			[styles.minimal.noneRounded]: !rounded,
		});
	return styles.defaultStates[variant];
};

const getButtonSize: (
	styles: typeof styleRefs,
	size: keyof typeof styles.size,
	rounded: boolean,
	iconOnly: boolean,
) => string = (styles, size, rounded, iconOnly) => {
	const currentSize = styles.size[size];

	return clsx(currentSize.default, {
		[currentSize.rounded]: rounded,
		[currentSize.iconOnly]: iconOnly,
	});
};
