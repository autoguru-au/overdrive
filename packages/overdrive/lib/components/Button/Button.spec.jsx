import { AccountBoxIcon } from '@autoguru/icons';
import { render } from '@testing-library/react';
import * as React from 'react';

import { Icon } from '../Icon';
import { Button } from './Button';

describe('<Button />', () => {
	it('should not throw', () =>
		expect(() => render(<Button />)).not.toThrow());

	it('should match snapshot for default button', () => {
		expect(render(<Button />).container.firstChild).toMatchSnapshot();
	});

	it.skip('should pass on className to dom element', () => {
		expect(
			render(<Button className="button-class" />).container.firstChild,
		).toHaveClass('button-class');
	});

	describe('when as button', () => {
		it('should use html button element by default', () => {
			expect(render(<Button />).container.firstChild).toMatchSnapshot();
		});
	});

	describe('when custom component', () => {
		it('should use html anchor element if href value exists', () => {
			expect(
				render(<Button is={<a href="abc" />} />).container.firstChild,
			).toMatchSnapshot();
		});

		it('should render passed in component', () => {
			expect(
				render(<Button is={<a href="/abcd" />}>Link button</Button>)
					.container.firstChild,
			).toHaveAttribute('href', '/abcd');
		});
	});

	describe('when colour variants', () => {
		it('should match snapshot for Primary button', () => {
			expect(
				render(<Button variant="primary" />).container.firstChild,
			).toMatchSnapshot();
		});

		it.skip('should apply Primary variant styles', () => {
			expect(
				render(<Button variant="primary" />).container.firstChild,
			).toHaveClass('variantPrimary');
		});

		it('should match snapshot for Secondary button', () => {
			expect(
				render(<Button variant="secondary" />).container.firstChild,
			).toMatchSnapshot();
		});

		it.skip('should apply Secondary variant styles', () => {
			expect(
				render(<Button variant="secondary" />).container.firstChild,
			).toHaveClass('variantSecondary');
		});

		it('should match snapshot for Danger button', () => {
			expect(
				render(<Button variant="danger" />).container.firstChild,
			).toMatchSnapshot();
		});

		it.skip('should apply Danger variant styles', () => {
			expect(
				render(<Button variant="danger" />).container.firstChild,
			).toHaveClass('variantDanger');
		});
	});

	describe('when size variant', () => {
		it('should match snapshot for small button', () => {
			expect(
				render(<Button size="small" />).container.firstChild,
			).toMatchSnapshot();
		});

		it.skip('should apply small size styles', () => {
			expect(
				render(<Button size="small" />).container.firstChild,
			).toHaveClass('small');
		});

		it('should match snapshot for medium button', () => {
			expect(
				render(<Button size="medium" />).container.firstChild,
			).toMatchSnapshot();
		});

		it.skip('should apply medium size styles', () => {
			expect(
				render(<Button size="medium" />).container.firstChild,
			).toHaveClass('medium');
		});
	});

	describe('when rounded', () => {
		it('should match snapshot', () => {
			expect(
				render(<Button rounded />).container.firstChild,
			).toMatchSnapshot();
		});

		it.skip('should not apply rounded styles to default button', () => {
			expect(render(<Button />).container.firstChild).not.toHaveClass(
				'rounded',
			);
		});

		it.skip('should apply rounded styles', () => {
			expect(render(<Button rounded />).container.firstChild).toHaveClass(
				'rounded',
			);
		});
	});

	describe('when isFullWidth', () => {
		it('should match snapshot', () => {
			expect(
				render(<Button isFullWidth />).container.firstChild,
			).toMatchSnapshot();
		});

		it.skip('should apply styles', () => {
			expect(
				render(<Button isFullWidth />).container.firstChild,
			).toHaveClass('fullWidth');
		});
	});

	describe('when loading', () => {
		it('should match snapshot for default button', () => {
			expect(
				render(<Button isLoading />).container.firstChild,
			).toMatchSnapshot();
		});

		it('should match snapshot for small button', () => {
			expect(
				render(<Button isLoading size="small" />).container.firstChild,
			).toMatchSnapshot();
		});

		it('should match snapshot for medium button', () => {
			expect(
				render(<Button isLoading size="medium" />).container.firstChild,
			).toMatchSnapshot();
		});

		it.skip('should not apply loading styles to default button', () => {
			expect(render(<Button />).container.firstChild).not.toHaveClass(
				'loading',
			);
		});

		it.skip('should apply loading styles', () => {
			expect(
				render(<Button isLoading />).container.firstChild,
			).toHaveClass('loading');
		});
	});

	describe.skip('when icon only', () => {
		it('should size to 20', () => {
			const { container } = render(
				<Button>
					<Icon icon={AccountBoxIcon} />
				</Button>,
			);

			expect(
				container.querySelectorAll('svg').item(0).parentNode.style
					.width,
			).toBe('20px');
		});
	});
});
