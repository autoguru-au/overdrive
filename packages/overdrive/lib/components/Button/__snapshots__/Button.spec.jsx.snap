// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot for default button 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base 4_mobile_base none_mobile_base 4_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base defaultStates_secondary_base variant_secondary_base size_medium_default_base enabled"
  type="button"
>
  <div
    class="base height_full center_mobile_base center_mobile_base body_base"
  />
</button>
`;

exports[`<Button /> when as button should use html button element by default 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base 4_mobile_base none_mobile_base 4_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base defaultStates_secondary_base variant_secondary_base size_medium_default_base enabled"
  type="button"
>
  <div
    class="base height_full center_mobile_base center_mobile_base body_base"
  />
</button>
`;

exports[`<Button /> when colour variants should match snapshot for Danger button 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base 4_mobile_base none_mobile_base 4_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base defaultStates_danger_base variant_danger_base size_medium_default_base enabled"
  type="button"
>
  <div
    class="base height_full center_mobile_base center_mobile_base body_base"
  />
</button>
`;

exports[`<Button /> when colour variants should match snapshot for Primary button 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base 4_mobile_base none_mobile_base 4_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base defaultStates_primary_base variant_primary_base size_medium_default_base enabled"
  type="button"
>
  <div
    class="base height_full center_mobile_base center_mobile_base body_base"
  />
</button>
`;

exports[`<Button /> when colour variants should match snapshot for Secondary button 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base 4_mobile_base none_mobile_base 4_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base defaultStates_secondary_base variant_secondary_base size_medium_default_base enabled"
  type="button"
>
  <div
    class="base height_full center_mobile_base center_mobile_base body_base"
  />
</button>
`;

exports[`<Button /> when custom component should use html anchor element if href value exists 1`] = `
<a
  class="base none_mobile_base 4_mobile_base none_mobile_base 4_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base defaultStates_secondary_base variant_secondary_base size_medium_default_base enabled"
  href="abc"
>
  <div
    class="base height_full center_mobile_base center_mobile_base body_base"
  />
</a>
`;

exports[`<Button /> when isFullWidth should match snapshot 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base 4_mobile_base none_mobile_base 4_mobile_base display_inlineBlock width_full overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base defaultStates_secondary_base variant_secondary_base size_medium_default_base enabled"
  type="button"
>
  <div
    class="base height_full center_mobile_base center_mobile_base body_base"
  />
</button>
`;

exports[`<Button /> when loading should match snapshot for default button 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base none_mobile_base none_mobile_base none_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base variant_secondary_base size_medium_default_base loading"
  disabled=""
  type="button"
>
  <div
    class="base size_medium_circular_base colours_secondary_base spinner"
  >
    <svg
      class="base overflow_hidden circular"
      viewBox="25 25 50 50"
    >
      <circle
        class="path_base size_medium_path_base"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
      />
    </svg>
  </div>
</button>
`;

exports[`<Button /> when loading should match snapshot for medium button 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base none_mobile_base none_mobile_base none_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base variant_secondary_base size_medium_default_base loading"
  disabled=""
  type="button"
>
  <div
    class="base size_medium_circular_base colours_secondary_base spinner"
  >
    <svg
      class="base overflow_hidden circular"
      viewBox="25 25 50 50"
    >
      <circle
        class="path_base size_medium_path_base"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
      />
    </svg>
  </div>
</button>
`;

exports[`<Button /> when loading should match snapshot for small button 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base none_mobile_base none_mobile_base none_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_3_base colours_white_base fontWeight_semiBold_base base variant_secondary_base size_small_default_base loading"
  disabled=""
  type="button"
>
  <div
    class="base size_medium_circular_base colours_secondary_base spinner"
  >
    <svg
      class="base overflow_hidden circular"
      viewBox="25 25 50 50"
    >
      <circle
        class="path_base size_medium_path_base"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
      />
    </svg>
  </div>
</button>
`;

exports[`<Button /> when rounded should match snapshot 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base 4_mobile_base none_mobile_base 4_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base pill_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base defaultStates_secondary_base variant_secondary_base size_medium_default_base size_medium_rounded_base enabled"
  type="button"
>
  <div
    class="base height_full center_mobile_base center_mobile_base body_base"
  />
</button>
`;

exports[`<Button /> when size variant should match snapshot for medium button 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base 4_mobile_base none_mobile_base 4_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_4_base colours_white_base fontWeight_semiBold_base base defaultStates_secondary_base variant_secondary_base size_medium_default_base enabled"
  type="button"
>
  <div
    class="base height_full center_mobile_base center_mobile_base body_base"
  />
</button>
`;

exports[`<Button /> when size variant should match snapshot for small button 1`] = `
<button
  class="base appearance cursorPointer button none_mobile_base 3_mobile_base none_mobile_base 3_mobile_base display_inlineBlock overflow_hidden textAlign_center border_style none_mobile_base none_mobile_base none_mobile_base none_mobile_base 1_mobile_base root_base root_base sizes_3_base colours_white_base fontWeight_semiBold_base base defaultStates_secondary_base variant_secondary_base size_small_default_base enabled"
  type="button"
>
  <div
    class="base height_full center_mobile_base center_mobile_base body_base"
  />
</button>
`;
